components/style/atomic_refcell.rs

```diff
@@ -2,121 +2,358 @@
  * License, v. 2.0. If a copy of the MPL was not distributed with this
  * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
 
+//! Implements a container type providing RefCell-like semantics for objects
+//! shared across threads.
+//!
+//! RwLock is traditionally considered to be the |Sync| analogue of RefCell.
+//! However, for consumers that can guarantee that they will never mutably
+//! borrow the contents concurrently with immutable borrows, an RwLock is
+//! overkill, and has key disadvantages:
+//! * Performance: Even the fastest existing implementation of RwLock (that of
+//!   parking_lot) performs at least two atomic operations during immutable
+//!   borrows. This makes mutable borrows significantly cheaper than immutable
+//!   borrows, leading to weird incentives when writing performance-critical
+//!   code.
```

Comment:
I don't think that's the issue - immutable borrows are ~twice as a slow as mutable borrows using parking_lot::RwLock. I haven't looked at the source closely, but I'm pretty sure there are twice as many atomic operations happening on read acquires.